docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+        - [NFR1](#nfr1)
+        - [NFR2](#nfr2)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.
+
+## Summary
+
+We propose a first-class Load Balancer construct for Cluster API for the purpose of providing a pluggable model for load balancing Kubernetes cluster control planes such that the load balancer implementation isn't tied directly into the Infrastructure Cluster resource. This will open the possibility for multiple swappable options in cloud environments and the ability to re-use load balancer implementations when possible, especially for bare metal and data center environments with heterogeneous infrastructure.
+
+## Motivation
+
+In private data center environments, there is no common infrastructure provider for load balancers vs. the provider for the machines themselves. Virtualized environments may use one vendor for virtual machines, e.g. VMware vSphere, KVM, or Xen, and a variety of vendors for networks and load balancers, e.g. F5, VMware NSX-T, Cisco ACI etc… This requires the ability to plug in different infrastructure providers for networking constructs with that of the machine provisioner. The most common use is the provisioning of a load balancer to sit in front of the Kubernetes control plane instances in order to provide a stable endpoint. Many cluster operators will also want to provision a stable machine-based load balancer to front all workload services from that provisioned cluster.
+
+Additionally, there are also cases where providers such as vSphere may be run in public cloud environments and being able to re-use existing native load balancer implementations would be preferred compared to having to re-implement or fork existing implementations for use in those environments.
+
+There are also cases within existing infrastructure provider implementations where it would be beneficial to allow for easier swapping of different load balancer implementations that may provide different benefits under different circumstances.
+
+We've also seen use cases where there is a need for a Load Balancer to front non-control plane services, such as Ingress. Any functionality that we add should also support these use cases in addition to just control plane load balancing.
```

Comment:
I think there currently is a clear separation between the two, mainly because of limitations of the existing CCM implementations. I covered a bit in the Alternative proposals section, but it could definitely use a bit more elaboration here, thanks!

Reply:
So in Azure there is actually direct overlap (not sure if the same applies for others), because each VM can only be associated with one LB type "internal" and one LB type "public", which means we have to reuse the same LB cloud provider creates for services of type `LoadBalancer` for things like node egress. See my comment here for more context: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1229#discussion_r599157066

> Is the idea that existing cloud providers can/should still be used in CAPI-managed clusters?

This should definitely be true, I wouldn't even suggest adding it as a goal

Reply:
Agree, there is a clear separation about what they can do.

I just think it's interesting how our (as CAPI project) view of the existing implementation is. Do we expect them to coexist or will our implementations supersede the existing ccm implementation (of course only in CAPI-managed clusters). Or don't we have an opinion and we don't need know as they can coexist without any effort on both sides. The overlap I currently see:
* we will depend on external CCMs so the Nodes get ready as soon as the in-tree implementation has been removed (although that's not the service controller, so it's a separate functionality in the CCM)
* we're currently also cleaning up CCM lbs during cluster deletion (at least in CAPA ([e2e-test](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/3181cf5a008dc19fbd62314d3cc16e14985675fa/test/e2e/suites/unmanaged/unmanaged_test.go#L207-L221), I also like to implement this in CAPO, but I don't know how other provider are handling this)
*

Reply:
this relates to this discussion: https://github.com/kubernetes-sigs/cluster-api/issues/3075

Reply:
@CecileRobertMichon thank you for the additional context. If I'm understanding correctly, this overlap already exists today, and I wouldn't expect anything in this proposal to greatly affect that overlap more than it exists currently.

I do want to better understand the use case, though, is it okay if I reach out next week to discuss things further over a higher bandwidth medium so that I can better understand the nuances involved and incorporate more of the Azure-specific requirements into the proposal?